### ISSUE TITLE
Update ubuntu 18.04 server modoboa 1.13.0 to 1.15.0 using installer failed uwsgi

### ISSUE THREAD

**Issue by demanl** [modoboa/modoboa-installer#328]
# Impacted versions

* Distribution:  Ubuntu 
* Codename: 18.04
* Arch: 64 Bits
* Database:  MySQL

# Steps to reproduce
downloaded and installed current modoboa installer.
copied original installer.cfg file to new modoboa-installer directory
ran "sudo ./run.py --upgrade dadllc.com"
ran through with no errors

root@mail2:/home/demanl/modoboa-installer# sudo ./run.py --upgrade dadllc.com
Welcome to Modoboa installer!

INSTALL dialog
Your mail server is about to be upgraded and the following components will be impacted:
modoboa automx amavis clamav dovecot nginx razor postfix postwhite spamassassin uwsgi radicale opendkim
Do you confirm? (Y/n) Y
The process can be long, feel free to take a coffee and come back later ;)
Starting...
Installing amavis
Installing spamassassin
Installing razor
Installing clamav
Installing modoboa
User modoboa already exists, skipping creation but please make sure the /srv/modoboa directory exists.
Installing automx
User automx already exists, skipping creation but please make sure the /srv/automx directory exists.
Installing radicale
User radicale already exists, skipping creation but please make sure the /srv/radicale directory exists.
Installing uwsgi
Installing nginx
Installing opendkim
Installing postfix
Installing postwhite
Installing dovecot
User vmail already exists, skipping creation but please make sure the /srv/vmail directory exists.
Congratulations! You can enjoy Modoboa at https://mail2.dadllc.com (admin:password)

ERROR frm log file
Current thread 0x00007fc1779155c0 (most recent call first):
Tue Jun 23 13:38:47 2020 - *** Starting uWSGI 2.0.15-debian (64bit) on [Tue Jun 23 13:38:47 2020] ***
Tue Jun 23 13:38:47 2020 - compiled with version: 7.3.0 on 28 September 2018 15:41:15
Tue Jun 23 13:38:47 2020 - os: Linux-4.15.0-106-generic #107-Ubuntu SMP Thu Jun 4 11:27:52 UTC 2020
Tue Jun 23 13:38:47 2020 - nodename: mail2.dadllc.com
Tue Jun 23 13:38:47 2020 - machine: x86_64
Tue Jun 23 13:38:47 2020 - clock source: unix
Tue Jun 23 13:38:47 2020 - pcre jit disabled
Tue Jun 23 13:38:47 2020 - detected number of CPU cores: 2
Tue Jun 23 13:38:47 2020 - current working directory: /
Tue Jun 23 13:38:47 2020 - writing pidfile to /run/uwsgi/app/modoboa_instance/pid
Tue Jun 23 13:38:47 2020 - detected binary path: /usr/bin/uwsgi-core
Tue Jun 23 13:38:47 2020 - dropping root privileges as early as possible
Tue Jun 23 13:38:47 2020 - setgid() to 1001
Tue Jun 23 13:38:47 2020 - setuid() to 1001
Tue Jun 23 13:38:47 2020 - chdir() to /srv/modoboa/instance
Tue Jun 23 13:38:47 2020 - your processes number limit is 15530
Tue Jun 23 13:38:47 2020 - your memory page size is 4096 bytes
Tue Jun 23 13:38:47 2020 - detected max file descriptor number: 1024
Tue Jun 23 13:38:47 2020 - VirtualHosting mode enabled.
Tue Jun 23 13:38:47 2020 - lock engine: pthread robust mutexes
Tue Jun 23 13:38:47 2020 - thunder lock: disabled (you can enable it with --thunder-lock)
Tue Jun 23 13:38:47 2020 - uwsgi socket 0 bound to UNIX address /run/uwsgi/app/modoboa_instance/socket fd 3
Tue Jun 23 13:38:47 2020 - dropping root privileges after socket binding
Tue Jun 23 13:38:47 2020 - Python version: 3.6.9 (default, Apr 18 2020, 01:56:04)  [GCC 8.4.0]
Tue Jun 23 13:38:47 2020 - Set PythonHome to /srv/modoboa/env
Fatal Python error: Py_Initialize: Unable to get the locale encoding
ModuleNotFoundError: No module named 'encodings

python --version returns Python 2.7   fyi

**Comment by demanl** [comment]
I'm now pretty sure that my problem was related to the fact that python 2 was dropped on May 7 2020.   And of course my install which occurred on Jan 2019 is using python 2.7 . :(   I did find a note about this under the 1.15.0 modoboa with a warning about having to rebuild my virtual environment.    Damn.     Now I need to learn how to rebuild a virtual environment.    A virtual environment in what ???   My head hurts.

**Comment by demanl** [comment]
Ok...I can update the virtual environment to python 3.6 pretty easily....after reading about python and finding link about rebuilding python virtual environments...pretty cool actually.  This fixes the uwsgi service.    

But...new problem....navigating to the webserver website now produces "Internal error" message in the browser.   Damn :(
I'm about to give up now.    All the documentation is written for modoboa developers.....not for any server administrator.   You have to be a seasoned python developer with several years of modoboa support/development time in order to make any sense out of their directions.    I'm pulling my hair out.    Thank god I can make snapshots of my server and revert back to them.   Seems like I'm stuck with running my original install of modoboa 1.13.0 :(

**Comment by tonioo** [comment]
@demanl Make sure to also update uwsgi config. You need the python3 plugin instead of python3. Check the /etc/uwsgi/apps-enabled/modoboa.ini, it should look like this one: https://github.com/modoboa/modoboa-installer/blob/master/modoboa_installer/scripts/files/uwsgi/modoboa.ini.tpl. Then, make sure the following packages are installed: https://github.com/modoboa/modoboa-installer/blob/master/modoboa_installer/scripts/uwsgi.py#L19.
Finaly, restart uwsgi.

**Comment by demanl** [comment]
root@mail2:/home/demanl# cat /etc/uwsgi/apps-enabled/modoboa_instance.ini
# This file was automatically installed on 2020-06-24T16:48:31.046198
[uwsgi]
uid = modoboa
gid = modoboa
plugins = python3
home = /srv/modoboa/env
chdir = /srv/modoboa/instance
module = instance.wsgi:application
master = true
processes = 2
vhost = true
no-default-app = true
socket = /run/uwsgi/app/modoboa_instance/socket
chmod-socket = 660
vacuum = true
single-interpreter = True

root@mail2:/home/demanl# dpkg --get-selections | grep uwsgi
uwsgi                                           install
uwsgi-core                                      install
uwsgi-plugin-python                             install
uwsgi-plugin-python3                            install

**Comment by demanl** [comment]
ok....I managed to > pip install -r requirements.txt  (built from a 1.13.0 env) and re-installed modoboa 1.13.0 in my python3.6 environment....and all seems to be working. 
I'm using a mysql database
And I'm STUCK at 1.13.0.        Both the installer-upgrade and manual install methods FAIL....:(     
I'm using ubuntu 18.04.    Is this the problem...???    What dist of linux are you running (exact distro) in development...???

**Comment by demanl** [comment]
ZOMG  I got Modoboa 1.15.0 and everything else that goes with INSTALLED and RUNNING on ubuntu 18.04.    
DAMN :)
I found the answer on your latest BLOG entry.
The key was I downloaded the latest modoboa-installer.
I copied my original installer.cfg file into the modoboa-installer directory.
Your blog mentioned that If I used the installer to install version using python2 and if I made no changes....I could just remove the virtual environment and install using the installer in upgrade mode!!!.    
Hmm.
So I deleted the /srv/modoboa/env/    virtual environment directory.....only.....leaving all the other /srv/modoboa directories intact.
I then cd  into my ...modoboa-installer/ directory....and ran "./run.py --upgrade <domain.com>" !!!!
And it ran....and it WORKED....holy cow.  :)
I then entered the virtual environment and checked version of python......and checked out the contents of /srv/modoboa/env/bin directory...and all is pointing to python3.6      :)
The uwsgi service was up and running....no issues.....I then went into the modoboa web admin website...and everything is working :)    All of the modules where updated to the latest.

I apologize for my lack of knowledge.......but the documentation does not come close to how easy it really is to update this product.

Cheers....:)
